components/skus/browser/skus_sdk_service.cc

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "brave/components/skus/browser/skus_sdk_service.h"
+
+#include <utility>
+
+#include "base/json/json_reader.h"
+#include "base/task/post_task.h"
+#include "base/threading/sequenced_task_runner_handle.h"
+#include "brave/components/skus/browser/rs/cxx/src/lib.rs.h"
+#include "brave/components/skus/browser/rs/cxx/src/shim.h"
+#include "brave/components/skus/browser/pref_names.h"
+#include "brave/components/skus/browser/skus_sdk_context_impl.h"
+#include "brave/components/skus/browser/skus_utils.h"
+#include "components/prefs/pref_service.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+
+namespace {
+
+void OnRefreshOrder(skus::RefreshOrderCallbackState* callback_state,
+                    skus::SkusResult result,
+                    rust::cxxbridge1::Str order) {
+  std::string order_str = static_cast<std::string>(order);
+  if (callback_state->cb) {
+    std::move(callback_state->cb).Run(order_str);
+  }
+  delete callback_state;
+}
+
+void OnFetchOrderCredentials(
+    skus::FetchOrderCredentialsCallbackState* callback_state,
+    skus::SkusResult result) {
+  if (callback_state->cb) {
+    std::move(callback_state->cb).Run("");
+  }
+  delete callback_state;
+}
+
+void OnPrepareCredentialsPresentation(
+    skus::PrepareCredentialsPresentationCallbackState* callback_state,
+    skus::SkusResult result,
+    rust::cxxbridge1::Str presentation) {
+  if (callback_state->cb) {
+    std::move(callback_state->cb).Run(static_cast<std::string>(presentation));
+  }
+  delete callback_state;
+}
+
+void OnCredentialSummary(skus::CredentialSummaryCallbackState* callback_state,
```

Comment:
I'd prefer these `rust-cxx` callbacks to be as thin as possible, i.e. can we bind all temporary items like `callback`, `domain`, `prefs`, etc. into `base::OnceCallback` and on rust response just do `callback_state.cb.Run(...)` with all stuff that we got from the rust side and `delete callback_state;` right after? This will require us to create `SkusSdkService::OnCredentialSummary` and similar methods, but this way it should be much harder to break/leak something accidentally. We can also use `WeakPtr`, which should protect us on shutdown.

Also, I see that `rust-cxx` does weird stuff like `callback_state.release()` on call hence the unexpected `delete` call requirement. Should we really use `unique_ptr`s then? I think raw pointers would be cleaner here, the WTF-ness will be applicable.

Reply:
currently we wait until the last possible moment before releasing the `unique_ptr`, as a result -  a cancelled rust future will properly clean up the C++ resource on `Drop`.

i'm working on a different way of structuring things that won't require us to use raw pointers at all though